### ISSUE TITLE
EventSetter. Check of HandlerType

### ISSUE THREAD

**Issue by lobster2012-user** [dotnet/wpf#806]
Hi!
(master) 
 **Problem description:**

``csharp
EventSetter.cs
if (_handler.GetType() != _event.HandlerType)
{
                throw new ArgumentException(SR.Get(SRID.HandlerTypeIllegal));
}

``
Why is there a strict check?
And if I want to set an action handler type <object, EventArgs>?
And there is such an opportunity (but not through styles)

[old question](https://social.msdn.microsoft.com/Forums/en-US/7b4ce02d-bdf1-4b7c-97d4-b7526b1792b2/wpfeventsetterstrong-handler-type?forum=wpf)

**Comment by weltkante** [comment]
Delegates don't have any inheritance hierarchy, the only sensible check you can do is a strict check. The `Invoke` method is generated by the compiler on the individual subclass, so while the Invoke method on `Action<object, EventArgs>` may have the same arguments as the Invoke method on `EventHandler` they are in no way related as far as both language and runtime are concerned.

Your example goes even further because you want to mix `Action<object, EventArgs>` with RoutedEventHandler, which Invoke method doesn't even have the same parameters.

While in theory these things make sense they'd require new features on every layer (the runtime, compiler and WPF framework level).

**Comment by lobster2012-user** [comment]
@weltkante 
Did not quite understand your answer.

https://github.com/lobster2012-user/WpfEventSetterSampleApp
```xml
<Button x:Name="ButtonClick" Click="ButtonClick_Click">Click</Button>
```
```csharp
private void ButtonClick_Click(object sender, EventArgs e)
```

Everything is already working, because it should work. 
The only problem is strict verification, which is not necessary.

**Comment by weltkante** [comment]
Your example actually works for me (and has nothing to do with strict checking quoted above).

In `MainWindow.g.cs` (in the obj folder of the project, generated by the xaml compiler) you have the line
```C#
this.ButtonClick.Click += new System.Windows.RoutedEventHandler(this.ButtonClick_Click);
```

And the delegate type correctly matches the delegate type required by the event. The compiler probably bridges the type mismatch of the arguments because it sees they are compatible.

**Comment by lobster2012-user** [comment]
I found my old project in which I ran into this problem.
I created the eventsetter with my hands and passed the delegate there not strictly type.
I updated the repo.
Two examples
1. Add handler via AddHandler
2. Create a manual eventsetter.

**Comment by weltkante** [comment]
Yes this is now an example which triggers the strict type check and fails. You are creating a `RoutedEventHandler` but the `MouseDownEvent` event needs a `MouseButtonEventHandler`.

However I already covered all relevant points in my previous response, you might want to reread and try to understand the points I'm bringing up:
* Delegates of different types are incomptible for the .NET runtime and WPF framework, only the C# language compiler can provide a little bit of assistance auto-converting delegates, but this doesn't happen in your case because you don't use the strongly typed event subscription method. As such the event handlers remain incompatible.
* Removing the typecheck just fails later when the event is raised with a typecast exception, unless you hit the very slow reflection based code path. Which code path you hit depends on the specific event are looking at, some subclasses of RoutedEventArgs (for example [this](https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/DragEventArgs.cs,193) one) will override the InvokeEventHandler call with a typecast to be more efficient than slow reflection based invocation.
* Removing the typecheck is a bad idea because it will make finding bugs really hard if you actually get the delegate wrong in an incompatible way

Considering that its a very simple fix in the caller to provide the correct event handler delegate, the disadvantages of not doing the check outweight the minor convenience you would get by removing the check.

**Comment by lobster2012-user** [comment]
I do not propose to remove checks, checks should be.

Inside routedEvent
there is less strict check
`` `csharp
internal bool IsLegalHandler (Delegate handler)
        {
            Type handlerType = handler.GetType ();
            
            return ((handlerType == HandlerType) ||
                     (handlerType == typeof (RoutedEventHandler)));
}
`` `
My question is why a condition was added to eventSetter that is stricter than that.

>>Delegates of different types are incomptible for the .NET runtime and WPF framework

**Here I do not agree**

``csharp

 public class MyArgs : EventArgs
    {

    }
    public class Tests
    {
        [SetUp]
        public void Setup()
        {
        }
        void OnEvent(object sender,EventArgs e)
        {
            Console.WriteLine(e.GetType().ToString());
        }
        [Test]
        public void Passed()
        {
            Delegate @delegate = new EventHandler(OnEvent);
            @delegate.DynamicInvoke(this, new MyArgs());
        }
        [Test]
        public void Failed()
        {
            Delegate @delegate = new EventHandler(OnEvent);
            Assert.Throws<ArgumentException>(() => @delegate.DynamicInvoke(this, 1));
        }
    }

```

**Comment by weltkante** [comment]
As far as I know `DynamicInvoke` is using reflection to look up the `Invoke` method on the delegate class on every single call, which is pretty slow compared to a normal method call. This is the "slow code path" I mentioned above. You probably don't want this for high frequency events (like mouse events, scrolling, drag'n'drop, etc.) which can be raised many times per second.

> I do not propose to remove checks, checks should be.
> Inside routedEvent there is less strict check

This check works because the [InvokeHandler](https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/RoutedEventHandlerInfo.cs,64) call of RoutedEventArgs has a special case for `RoutedEventHandler`

It may be safe to replace [the strict check in EventSetter](https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/EventSetter.cs,114) by a call to [IsLegalHandler](https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/RoutedEvent.cs,83) - I'll leave it to someone from the WPF team to double check the technical details.

In any case it just adds `RoutedEventHandler` as additional option and won't help you calling something via `Action<object, EventArgs>` as you requested in your initial issue description.

**Comment by lobster2012-user** [comment]
>>In any case it just adds RoutedEventHandler as additional option and won't help you calling something via Action<object, EventArgs> as you requested in your initial issue description.

I replaced ~~EventArgs~~ with **RoutedEventArgs** in the first message to leave compatibility.

>>It may be safe to replace the strict check in EventSetter by a call to IsLegalHandler - I'll leave it to someone from the WPF team to double check the technical details.

YES.


About reflection and speed. 
Here I agree.
With a great desire this is all solved by runtime code generation. 
But that's another story. In that project, I had to do it.
